keps/sig-node/NNNN-memory-manager-hugepages-verification/README.md

```diff
@@ -0,0 +1,478 @@
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes enhancing the Memory Manager's Static policy to verify OS-reported
+free hugepages availability during pod admission. Currently, the Memory Manager only
+tracks hugepage allocations for Guaranteed QoS pods but doesn't verify actual
+hugepage availability from the operating system. This can lead to pods being admitted
+when hugepages aren't actually available, causing runtime failures.
+
+The enhancement adds verification by reading free hugepages from sysfs
+(`/sys/devices/system/node/node<N>/hugepages/hugepages-<size>kB/free_hugepages`)
+during pod admission, ensuring pods requesting hugepages are only admitted when
+sufficient free hugepages exist.
+
+## Motivation
+
+The Memory Manager tracks hugepage allocations for Guaranteed QoS pods to provide
+NUMA-aware memory and hugepage pinning. However, it operates on its internal
+accounting without verifying the actual state of hugepages on the system.
+
+This creates a problem when:
+1. Burstable or BestEffort pods consume hugepages (via hugetlbfs mounts or
+   `mmap` with `MAP_HUGETLB`) without being tracked by the Memory Manager
+2. External processes or other system components consume hugepages
```

Comment:
One of the key assumptions of how the kubelet operates in general, is that is the sole owner of a node. There is some leeway in some cases: we can pre-partition CPUs and make the kubelet assume it is the sole owner of _the resource pool it got when started_, and we can probably should do the same for hugepages. But in general, dynamic co-sharing of resources (kubelet races with other daemons or programs) is not supported and it's unlikely it ever will